### PR TITLE
Add simulation plugins, PDK class and components

### PR DESCRIPTION
**Pull request by SkandanC**
Major changes:

- Added PDK class
- Added Lumerical simulation plugins
- Added mzi

Still some work to be done, have to add some documentation.

### CODE REVIEW

**Review by sebastian-goeldi — CHANGES_REQUESTED** [review]

File: src/kfactory/kcell.py [hunk 1]
```diff
@@ -1404,6 +1404,9 @@ def to_yaml(cls, representer, node):  # type: ignore
         if shapes:
             d["shapes"] = shapes
         if len(node.settings) > 0:
+            for key, value in node.settings.items():
```
Comment (sebastian-goeldi) [inline src/kfactory/kcell.py:1407]: Why is this necessary? When does it make sense to have a callable setting?

File: src/kfactory/pcells/DCs.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    sbend_2 = c << bend_s(
+        width=width,
+        height=((dy) / 2 - gap / 2 - width / 2),
+        length=dx,
+        layer=layer,
+        enclosure=enclosure,
+    )
+
+    sbend.transform(kdb.DTrans(0, True, 0, (gap - width + dy)))
+    sbend_2.transform(kdb.DTrans(2, False, 0, -(gap + width + dy)))
```
Comment (sebastian-goeldi) [inline src/kfactory/pcells/DCs.py:58]: Transforming them to then connect them? That shouldn't be necessary

File: src/kfactory/pcells/__init__.py [hunk 3]
```diff
@@ -1,2 +1,28 @@
...
+]
+
+pcells: dict[str, Callable[..., kf.kcell.KCell]] = {  # type: ignore
+    "bend_s": bend_s,
+    "bend_circular": bend_circular,
+    "bend_euler": bend_euler,
+    "bend_s_euler": bend_s_euler,
+    "taper": taper_function,
+    "waveguide": wg,
+}
```
Comment (sebastian-goeldi) [inline src/kfactory/pcells/__init__.py:28]: why?

File: src/kfactory/pcells/DCs.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            "component": "straight_coupler",
+            "params": {
+                "gap": gap,
+                "length": length,
+                "width": width,
+                "layer": layer,
+                "enclosure": enclosure,
+            },
+        },
+    }
```
Comment (sebastian-goeldi) [inline src/kfactory/pcells/DCs.py:114]: Why add info that the cell already knows?

File: src/kfactory/pcells/dbu/waveguide.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 from typing import Optional
 
 from ... import KCell, LayerEnum, autocell, kdb, klib
+from ...pdk import _ACTIVE_PDK
```
Comment (sebastian-goeldi) [inline src/kfactory/pcells/dbu/waveguide.py:4]: Why? The waveguide doesn't care about the pdk. It can use it, but it does not depend on it

**Review by sebastian-goeldi — CHANGES_REQUESTED** [review]

File: src/kfactory/pdk.py [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+    Parameters:
+        name: PDK name.
+        enclosures: dict of enclosures factories.
+        cells: dict of parametric cells that return Components.
+        symbols: dict of symbols names to functions.
+        default_symbol_factory:
+        containers: dict of pcells that contain other cells.
+        base_pdk: a pdk to copy from and extend.
+        default_decorator: decorate all cells, if not otherwise defined on the cell.
+        layers: maps name to gdslayer/datatype.
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/pdk.py:51]: I set it to be a dict so the user can define it as {LAYER.xyz.name: LAYER.xyz}

File: src/kfactory/pdk.py [hunk 7]
```diff
@@ -0,0 +1,738 @@
...
+            For example dict(si=(1, 0), sin=(34, 0)).
+        layer_stack: maps name to layer numbers, thickness, zmin, sidewall_angle.
+            if can also contain material properties
+            (refractive index, nonlinear coefficient, sheet resistance ...).
+        layer_views: includes layer name to color, opacity and pattern.
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/pdk.py]: Removed that attribute.

File: src/kfactory/simulation/get_sparameters_path.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+from copy import deepcopy
+from functools import partial
+from pathlib import Path
+from typing import Optional
+
+import numpy as np
+
+import kfactory as kf
+from ..generic_tech import LayerStack
+from gdsfactory.name import clean_value
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/simulation/get_sparameters_path.py]: Set it to from kfactory.kcell

File: src/kfactory/pcells/dbu/waveguide.py [hunk 9]
```diff
@@ -10,11 +11,12 @@
 def waveguide(
     width: int,
     length: int,
-    layer: int | LayerEnum,
+    layer: int | LayerEnum | str,
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/pcells/dbu/waveguide.py]: Done

File: src/kfactory/pcells/bezier.py [hunk 10]
```diff
@@ -46,13 +46,13 @@ def bend_s(
     if enclosure is None:
         enclosure = Enclosure()
 
-    enclosure.extrude_path(c, path=pts, main_layer=layer, width=width)
+    enclosure.extrude_path(c, path=pts, main_layer=layer, width=width, end_angle=0)
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/pcells/bezier.py]: Done

File: src/kfactory/kcell.py [hunk 11]
```diff
@@ -2666,7 +2666,7 @@ def show(
         tf.parent.mkdir(parents=True, exist_ok=True)
         gds.write(str(tf), save_options)
         gds_file = tf
-        delete = True
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/kcell.py]: Done

File: src/kfactory/components/gc.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
+import kfactory as kf
+from kfactory.pcells import grating_coupler_elliptical
+from kfactory.tech import LAYER
+
+
+@kf.autocell
+def GC1() -> kf.KCell:
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/components/gc.py]: Done

File: src/kfactory/components/DCs.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+    wg_bottom = c << waveguide(width, length, layer, enclosure)
+    wg_bottom.trans = kdb.DTrans(0, False, 0, -(gap + width) / 2)
+
+    c.add_port(name="o1", port=wg_top.ports["o1"])
+    c.add_port(name="o2", port=wg_top.ports["o2"])
+    c.add_port(name="o3", port=wg_bottom.ports["o2"])
+    c.add_port(name="o4", port=wg_bottom.ports["o1"])
+
+    c.draw_ports()
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/components/DCs.py]: Removed, forgot to remove it after debugging 😃

File: src/kfactory/kcell.py (hunk 1 above)
Comment (SkandanC) [reply src/kfactory/kcell.py:1407]: See for instance the mzi pcell, where some arguments are functions to other pcells. Perhaps the arguments should always be kcells?

File: src/kfactory/pcells/DCs.py (hunk 2 above)
Comment (SkandanC) [reply src/kfactory/pcells/DCs.py:58]: Edited

File: src/kfactory/pcells/DCs.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+        length=dx,
+        layer=layer,
+        enclosure=enclosure,
+    )
+
+    sbend_r_top.connect("E0", wg.ports["o2"])
+    sbend_r_bottom.connect("E0", wg.ports["o3"])
+
+    sbend_r_top.transform(kdb.DTrans(0, False, 0, -(gap + width)))
+    sbend_r_bottom.transform(kdb.DTrans(0, False, 0, (gap + width)))
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/pcells/DCs.py:83]: Edited

File: src/kfactory/pcells/__init__.py (hunk 3 above)
Comment (SkandanC) [reply src/kfactory/pcells/__init__.py:28]: This is used by get_component to get KCells from the .pcells

File: src/kfactory/pcells/DCs.py (hunk 4 above)
Comment (SkandanC) [reply src/kfactory/pcells/DCs.py:114]: Edited

File: src/kfactory/pcells/dbu/waveguide.py (hunk 5 above)
Comment (SkandanC) [reply src/kfactory/pcells/dbu/waveguide.py:4]: Edited

File: src/kfactory/utils/geo.py [hunk 15]
```diff
@@ -665,12 +665,22 @@ def extrude_path(
         path: list[kdb.DPoint],
         main_layer: Optional[int | LayerEnum],
         width: float,
+        start_angle: float = 0,
+        end_angle: float = 0,
```
[the remark this answers is not in the document]
Comment (SkandanC) [reply src/kfactory/utils/geo.py:669]: Done

File: src/kfactory/simulation/write_sparameters_lumerical.py [hunk 16]
```diff
@@ -0,0 +1,1076 @@
...
+                component_,
+                dirpath=dirpath,
+                layer_stack=layer_stack or get_layer_stack(),
+                **settings,
+            )
+    if len(s_params) > 1:
+        try:
+            import sys
+
+            sys.path.append("C:\\Program Files\\Lumerical\\v231\\api\\python\\")
```
Comment (joamatab) [inline src/kfactory/simulation/write_sparameters_lumerical.py:265]: remove this?
